rfcs/0010-custom-system-uris.md

```diff
@@ -0,0 +1,325 @@
+From the perspective of their business decision makers,
+  the only thing that really matters is that these URIs are stable:
+  that we do not change our API in ways that are backwards-incompatible with their existing applications.
+That's about the sum total of our users' concerns:
+
+* Ensure that our URIs are documented,
+    such that they can be copy-pasted into their application code.
+* Ensure that our URIs are descriptive,
+    such that engineers building applications against our APIs can make reasonable guesses as to their meaning.
+* Ensure that our APIs continue to work with their applications.
+
+As a bonus, it's helpful if our URIs are also URLs for documentation relating to themselves,
+  such that engineers can copy-paste them into their browsers and get detailed information on the field they represent/identify.
+
+Those are our first-order concerns.
+Internally, we of course have other goals.
+
+Primary among those is a natural desire for our URLs to match our product and/or organizational structure.
+For example, if the Awesome FHIR API is being offered by the Office of Fun,
+  in the Entertaining Parties (EP) agency,
+  which is a part of the Department of Population Happiness (DPH),
+  which is part of the UK government,
+  a base URI like the following might make sense:
+  `https://awesomefhir.fun.ep.dph.gov.uk/`.
+Or variants of it.
+For BFD, that could look something like this:
+  `https://bfd.dasg.oeda.cms.gov/`.
+Or variants of it.
+
+
+## Proposed Solution
+[Proposed Solution]: #proposed-solution
+
+<!--
+Explain the proposal as if it was already implemented and shipped, and you were just explaining it to another developer or user.
+That generally means:
+
+* Introducing new named concepts.
+* Identifying and address each of the various audiences who might (or should) care about this proposal.
+  Explaining the solution using concepts and terms relevant to eaach of them.
+  Explaining how they should _think_ about the solution; detailing the impact as concretely as possible.
+  Possible audiences might include:
+    * Internal team: engineers, operators, product management, business owners.
+    * External users: engineers, operators, product management, business owners, end users.
+* Explaining the feature largely in terms of examples.
+    * Screencasts are often a good idea.
+        * On Mac OS X, you can use the built-in Quicktime Player or the built-in Mac OS X Mojave (and up) feature.
+    * Diagrams are often a good idea.
+        * Keep it simple! Use something like <http://asciiflow.com/>.
+* As part of implementing this proposal, will any documentation updates be needed, e.g. changelogs, Confluence pages, etc.?
+  If so, draft them now! Include the draft as a subsection or addendum.
+-->
+
+This RFC makes the recommendation that BFD should stick with a base of
+  `https://bluebutton.cms.gov/resources/` for all custom URIs,
+  _at least_ until a v3 of the API.
```

Comment:
Should we not 'bake in' versioning? i.e., https://bluebutton.cms.gov/resources/v1/. Allows us to be both backwards compatible as well as providing an avenue of future changes.

Reply:
In this particular case, I think I'd lean more towards pushing the versions down, e.g. `https://bluebutton.cms.gov/resources/variables/bene_id/v2` or something vaguely like that. Most codesystems won't change from release to release, so it'd be nice to avoid forcing a bunch of unnecessary app changes with new API releases.

Thoughts?

Reply:
I like that idea better; also, aligns more with the strategy where future functionality may be added to say R4 services, but not R2 (except in the case of an actual bug in R2). So level-setting at R4 but allowing say a /v1 to map to our v1 functionality as needed.

Reply:
The only concern on this front would be contention with the API versioning (which is at a higher level) and the ability to know if on the most current component.   Would there be some sort of map along with this lower chain versioning.   It is somewhat non-intuitive to be using a /v4 api with /v1 set of resources.   Though likely way off the map here.   I liked the higher level /version component due to this, though can see additional work to keep updated, and the developer inference that updates are constantly needed.   Not against the lower level versioning, though had this concern.

Reply:
I'll add one comment here.  I would recommend we stick with the status quo for now. When we release a V3 of the API,  I think there is a bigger discussion about properly defining valuesets, codesystems and extensions.  Currently these URLs (e.g. [https://bluebutton.cms.gov/resources/variables/](https://bluebutton.cms.gov/resources/variables/))  are really just webpages.  Ideally they would point at actual  ValueSet and CodeSystem definitions - likewise extensions should point to the corresponding StructureDefinition.   I realize this would be a breaking change, thus a V3 type item., but it would allow the Validator to catch valueset errors.   Currently these are basically ignored.